src/libpmem2/pmem2_windows.c

```diff
@@ -35,71 +35,109 @@
  */
 
 #include "libpmem2.h"
+#include "out.h"
 #include "pmem2.h"
 #include "pmem2_utils.h"
 
+#define PMEM2_HIDWORD(x) ((DWORD)((x) >> 32))
+#define PMEM2_LODWORD(x) ((DWORD)((x) & 0xFFFFFFFF))
+
 /*
- * pmem2_map -- XXX
+ * mapping -- creates file mapping object for a specified file
  */
-int
-pmem2_map(struct pmem2_config *cfg, struct pmem2_map **mapp)
+HANDLE
+mapping(const struct pmem2_config *cfg, DWORD protect, int *err)
 {
-	int ret = PMEM2_E_OK;
-	struct pmem2_map *map;
-	map = (struct pmem2_map *)pmem2_zalloc(&ret, sizeof(*map));
-	if (!map)
-		ret;
-	map->cfg = (struct pmem2_config *)pmem2_zalloc(&ret, sizeof(*map->cfg));
-	if (!map->cfg) {
-		goto err_free_map;
-	}
-	memcpy(map->cfg, cfg, sizeof(*cfg));
-	map->addr = NULL;
-	*mapp = map;
-
-	DWORD access = FILE_MAP_ALL_ACCESS;
 	size_t max_size = cfg->length + cfg->offset;
-	HANDLE fh = CreateFileMapping(cfg->handle,
+	*err = 0;
+	HANDLE mh = CreateFileMapping(cfg->handle,
 		NULL, /* security attributes */
-		PAGE_READWRITE,
-		(DWORD)(max_size >> 32),
-		(DWORD)(max_size & 0xFFFFFFFF),
+		protect,
+		PMEM2_HIDWORD(max_size),
+		PMEM2_LODWORD(max_size),
 		NULL);
 
-	if (!(fh) || fh == INVALID_HANDLE_VALUE) {
-		ret = GetLastError();
-		if (ret == ERROR_ACCESS_DENIED) {
-			fh = CreateFileMapping(cfg->handle,
-				NULL, /* security attributes */
-				PAGE_READONLY,
-				(DWORD)(max_size >> 32),
-				(DWORD)(max_size & 0xFFFFFFFF),
-				NULL);
-			access = FILE_MAP_READ;
-		}
-		if (!(fh) || fh == INVALID_HANDLE_VALUE) {
-			ret = PMEM2_E_INVALID_HANDLE;
-			goto err_free_map_cfg;
-		}
+	*err = GetLastError();
+	if (!mh) {
+		ERR("CreateFileMapping, error: 0x%08x", err);
+	} else if (*err == ERROR_ALREADY_EXISTS) {
+		ERR("CreateFileMapping, error: 0x%08x", err);
+	}
+
+	return mh;
+}
+
+/*
+ * pmem2_map -- map memory according to provided config
+ */
+int
+pmem2_map(const struct pmem2_config *cfg, struct pmem2_map **mapp)
+{
+	int ret = PMEM2_E_OK;
+
+	DWORD access = FILE_MAP_ALL_ACCESS;
+	HANDLE mh = mapping(cfg, PAGE_READWRITE, &ret);
+	if (ret == ERROR_ACCESS_DENIED) {
+		mh = mapping(cfg, PAGE_READONLY, &ret);
+		access = FILE_MAP_READ;
+	}
+	if (ret == ERROR_ALREADY_EXISTS) {
+		ret = PMEM2_E_MAPPING_EXISTS;
+		goto err;
+	}
+	if (!mh) {
+		ret = PMEM2_E_EXTERNAL;
+		return ret;
 	}
```

Comment:
I suggest checking for the empty mh first. Otherwise, you can try to close a NULLed handle.